### DISCUSSION TITLE
regex on micropython 1.19 using \w

### DISCUSSION THREAD

**Discussion by privatefcagney** [micropython/micropython#9741] [General]
The documentation is pretty clear. \w is supported.

But this doesn't work. No re.compile error. It just doesn't find.
    `rDirchar = '[\\w\\._]+'`

This however does work:
    `rDirchar = `'[A-Za-z0-9\\._]+'`

Note I specifically didn't use r strings i.e. r'\w'

At first I though my version was too old but I'm on a pretty recent one:
`MicroPython v1.19.1-451-gbdbc44474 on 2022-09-27; linux [GCC 11.2.0] version`

BTW is this even the right place to make such a post. I have a big banner warning me this is for orginization level discussions. But there's also lots of "dumb questions" like mine.

**Comment by privatefcagney** [comment]
NOTE There are two backslash before the dot in the working regex. 
page formatting removed one of my backslash rending my code stupid. Unescaped . is almost anything

**Comment by karfas** [comment]
While I have ho idea what your first expression should match (`\w` includes the underscore `_`, there are too many '\\' in the string and there is no need to escape the dot inside of character classes), I assume that character classes inside of character classes is one of the regex constructs not supported in micropython.
You might want to check https://docs.micropython.org/en/latest/library/re.html - this lists some unsupported things and **etc**.
In standard python, I would write your expression as
```
>>> re.compile('[\w.]+').match('foo_bar.xx')
<re.Match object; span=(0, 10), match='foo_bar.xx'>
>>> re.compile('[\w.]+').match('foo_#bar.xx')
<re.Match object; span=(0, 4), match='foo_'>
```

**Reply by privatefcagney** [reply]
You are right \w does indeed include _. `[\w_]` was not wrong, just the `_` was redundant.
I also didn't realise there's no need to cast . inside []

There are however NOT too many \ inside the string. 
You either need to use rstrings which treat \ as a normal character or cast \ with itself.

However with this new found knowledge I still cannot get \w or . to work

```
Python 3.10.6 (main, Aug 10 2022, 11:40:04) [GCC 11.3.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import re
>>> rx = re.compile('[\\w.]+')
>>> dut = 'my_file.py'
>>> m = rx.match(dut)
>>> m.group(0)
'my_file.py'

```
compared to micropython:

```
MicroPython v1.19.1-451-gbdbc44474 on 2022-09-27; linux [GCC 11.2.0] version
Use Ctrl-D to exit, Ctrl-E for paste mode
>>> import re
>>> rx = re.compile('\\w+')
>>> dut = 'my_file.py'
>>> m = rx.match(dut)
>>> m.group(0)
'my_file'
>>> rx = re.compile('[\\w.]+')
>>> m = rx.match(dut)
>>> m.group(0)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AttributeError: 'NoneType' object has no attribute 'group'

```
\w is equivalent to [A-Za-z0-9_] so my guess is micropython is parsing 
`[\\w.]` as `[[A-Za-z0-9_].]`
This is most certainly not documented. I carefully read the documentation before posting here.

**Reply by karfas** [reply]
> There are however NOT too many \ inside the string. You either need to use rstrings which treat \ as a normal character or cast \ with itself.

Well, escaping all `\` in a regex might be good practice in python (even when the regex becomes unreadable), but is not strictly a **requirement** with the `\w` example - `\w` works in a regex with or without escaped backslash. 
Inside of strings, only a few characters like like `\t`, `\n` get interpreted by python - for all others, python happily uses the single(!) backslash found.
`'\w'` => backslash followed by 'w'
`'\n'` => a newline character
```
>>> len('\w')
2
>>> len('\n')
1
``` 
I have no idea how a micropython regex treats `[\w]` internally, but obviously it is not working...

**Reply by privatefcagney** [reply]
Karfas I owe you thanks and an apology.

Seems I've misunderstood casting chars in strings for decades. Not a big problem because overcasting isn't wrong, just less readable.
In particular I though \ was always interpreted. 

Seems instead if the compiler recognises it as a cast: \\ \n \r \0 etc it performs the cast
If it doesn't recognise it: \w \s \[ those characters are reproduced exactly as is, which is what we want for regex strings.

micropython regex does understand (\w|\.), which under cpython could be: [\w.] 
But micropython doesn't handle [\w.]

**Reply by karfas** [reply]
> Karfas I owe you thanks and an apology.

No apology required. I am still learning python and I'm pretty sure that this  is an language feature/inconsistency I will hate in the long run, as it's different to many other languages (C, perl) I already know.

> micropython regex does understand (\w|.), which under cpython could be: [\w.]

Nitpicking: It should be `(\w|\.)` to get similar behaviour like `[\w.]`. The expression `(\w|.)` matches any character (due to `.`).

With CPython:
```
>>> import re
>>> re.compile('(\w|.)').match('#')
<_sre.SRE_Match object at 0x7fe8c5759a80>
```

**Reply by privatefcagney** [reply]
Nitpicking! Actually this time it was a typo rather than ignorance
